src/term/mod.rs

```diff
@@ -613,67 +617,74 @@ impl ansi::Handler for Term {
     #[inline]
     fn input(&mut self, c: char) {
 
-        if self.cursor.col == self.grid.num_cols() {
+        if self.wrap {
+            
             debug_println!("wrapping");
+        
             {
                 let location = Point {
                     line: self.cursor.line,
-                    col: self.cursor.col - 1
+                    col: self.cursor.col
                 };
 
                 let cell = &mut self.grid[&location];
                 cell.flags.insert(cell::WRAPLINE);
             }
+           
             if (self.cursor.line + 1) >= self.scroll_region.end {
                 self.linefeed();
             } else {
                 self.cursor.line += 1;
             }
+           
             self.cursor.col = Column(0);
+            self.wrap = false;
         }
 
-        unsafe {
-            if ::util::unlikely(self.cursor.line == self.grid.num_lines()) {
-                panic!("cursor fell off grid");
-            }
+        {
+            let cell = &mut self.grid[&self.cursor];
+            *cell = self.template_cell;
+            cell.c = c;
+        } 
+        
+        if (self.cursor.col + 1) < self.grid.num_cols() {
+            self.cursor.col += 1;
+        } else {
+            self.wrap = true;
         }
 
-        let cell = &mut self.grid[&self.cursor];
-        *cell = self.template_cell;
-        cell.c = c;
-        self.cursor.col += 1;
     }
 
     #[inline]
     fn goto(&mut self, line: Line, col: Column) {
-        use std::cmp::min;
         debug_println!("goto: line={}, col={}", line, col);
-        self.cursor.line = min(line, self.grid.num_lines() - 1);
-        self.cursor.col = min(col, self.grid.num_cols() - 1);
+        self.cursor.line = cmp::min(line, self.grid.num_lines() - 1);
+        self.cursor.col = cmp::min(col, self.grid.num_cols() - 1);
+        self.wrap = false;
     }
 
     #[inline]
     fn goto_line(&mut self, line: Line) {
-        use std::cmp::min;
         debug_println!("goto_line: {}", line);
-        self.cursor.line = min(line, self.grid.num_lines() - 1);
+        self.cursor.line = cmp::min(line, self.grid.num_lines() - 1);
```

Comment:
I was trying not to change the style too much. A previous PR imported std::cmp::min and used it in two different functions. When i went to import it at the module level I had found that std::cmp was already imported and there were a few uses of "::cmp::min" and "cmp::min" I did not know what style you were so I tried not to change everything, but thought they should be uniform at least. That being said I can go import std::cmp::min and make sure all the cases use just min.